sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/PubsubMessageToRow.java

```diff
@@ -68,8 +69,21 @@
 
   public abstract boolean useDlq();
 
+  public abstract boolean useFlatSchema();
+
   private Schema payloadSchema() {
-    return messageSchema().getField(PAYLOAD_FIELD).getType().getRowSchema();
+    if (useFlatSchema()) {
+      Schema.Builder builder = Schema.builder();
+      for (Schema.Field field : messageSchema().getFields()) {
```

Comment:
nit: might be a pithy way to do e.g. `messageSchema().getFields().stream().filter(f -> !f.getName().equals(TIMESTAMP_FIELD)` but it might just get crufty anyhow

Reply:
Done. I'm a fan of pithy things with streams